text/3316-array-vec.md

```diff
@@ -0,0 +1,407 @@
+  internally. This function wraps a `Guard` struct containing a `MaybeUninit`
+  array and a length that it initializes item by item. Essentially, _this is the
+  fundamental structure of `ArrayVec`_, it is just not made public. Having
+  `ArrayVec` would allow simplifying this function and others like it.
+- The much-requested feature of some way to collect into arrays would have a
+  more clear path, potentially by making `try_collect_into_array` public
+- Constructing a `core::ffi::CStr` is not directly possible from `&str` due to
+  the extra bit needed. `ArrayVec` would allow for a more clear way to perform
+  this common operation in `no_std` environments.
+- A structure such as `ArrayString` would be posssible to enable easier string
+  manipulation in `no_std` environments
+
+In short, the benefits to an `ArrayVec` concept are notable enough that there
+are already parts of the implementation in core, and there are a handful of top
+100 crates that provide similar functionality. Exporsing a public `ArrayVec` in
+`core` would help reduce fragmentation, provide a pathway for future language
+features, and give users a builtin tool for a common form of data manipulation.
+
+
+[try_collect_arr]: https://github.com/rust-lang/rust/blob/17cbdfd07178349d0a3cecb8e7dde8f915666ced/library/core/src/array/mod.rs#L804
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`ArrayVec` is a simple data structure, represented internally with a fixed-size
+memory buffer (an array) and a length. It should feel very familiar to `Vec`.
+The main difference to `Vec` is, the maximum capacity of that memory buffer must
+be known at compile time, and is specified through a generic paramter. See the
+comments in this example:
+
+```rust
+use core::collections::ArrayVec;
+
+const ARR_LEN: usize = 4;
+
+// We are creating an `ArrayVec` here that contains instances of i32, and has a capacity
+// of 4. That capacity cannot be changed during runtime
+let mut v: ArrayVec<i32, ARR_LEN> = ArrayVec::new();
+
+// Adding values to this `ArrayVec` works almost as you would expect
+// One difference is, `push()` returns a `Result<(), InsertionError>`.
+// This is because there is a much higher chance that the insertion may fail at
+// runtime (running out of space on the buffer) compared to `Vec`
+v.push(1).unwrap();
+v.push(2).unwrap();
+
+// Length and indexing work similarly to other data structures
+assert_eq!(v.len(), 2);
+assert_eq!(v[0], 1);
+assert_eq!(v.pop(), Some(2));
+assert_eq!(v.len(), 1);
+
+// Indexed assignment works as expected
+// **RFC TODO** what is the safe/checked way to perform assignment? It seems
+// like there should be a `.set(&T) -> Result` method to go along with `get`,
+// but I don't know what it is (probably just missing something)
+v[0] = 7;
+assert_eq!(v[0], 7);
+
+// Many higher-order concepts from `Vec` work as well
+v.extend([1, 2, 3].iter().copied());
```

Comment:
This is an interesting one, since it has the same "panic vs truncate vs …" question that collecting has.

It reminds me of the discussions in https://github.com/rust-lang/rfcs/pull/3271#issuecomment-1193389495 -- maybe there could be a similar API for "extend or push or … within existing capacity" between normal `Vec` *and* `ArrayVec`.